java/ql/lib/semmle/code/java/security/ImplicitPendingIntents.qll

```diff
@@ -0,0 +1,125 @@
+private import semmle.code.java.dataflow.TaintTracking
+private import semmle.code.java.frameworks.android.Intent
+private import semmle.code.java.frameworks.android.PendingIntent
+
+/** A source for an implicit `PendingIntent` flow. */
+abstract class ImplicitPendingIntentSource extends DataFlow::Node {
+  /** Holds if this source has the specified `state`. */
+  predicate hasState(DataFlow::FlowState state) { state = "" }
+}
+
+/** A sink that sends an implicit and mutable `PendingIntent` to a third party. */
+abstract class ImplicitPendingIntentSink extends DataFlow::Node {
+  /** Holds if this sink has the specified `state`. */
+  predicate hasState(DataFlow::FlowState state) { state = "" }
+}
+
+/**
+ * A unit class for adding additional taint steps.
+ *
+ * Extend this class to add additional taint steps that should apply to flows related to the use
+ * of implicit `PendingIntent`s.
+ */
+class ImplicitPendingIntentAdditionalTaintStep extends Unit {
+  /**
+   * Holds if the step from `node1` to `node2` should be considered a taint
+   * step for flows related to the use of implicit `PendingIntent`s.
+   */
+  predicate step(DataFlow::Node node1, DataFlow::Node node2) { none() }
+
+  /**
+   * Holds if the step from `node1` to `node2` should be considered a taint
+   * step for flows related to the use of implicit `PendingIntent`s. This step is only applicable
+   * in `state1` and updates the flow state to `state2`.
+   */
+  predicate step(
+    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2,
+    DataFlow::FlowState state2
+  ) {
+    none()
+  }
+}
+
+private class IntentCreationSource extends ImplicitPendingIntentSource {
+  IntentCreationSource() {
+    exists(ClassInstanceExpr cc |
+      cc.getConstructedType() instanceof TypeIntent and this.asExpr() = cc
+    )
+  }
+}
+
+private class SendPendingIntent extends ImplicitPendingIntentSink {
+  SendPendingIntent() {
+    sinkNode(this, "intent-start") and
+    // implicit intents can't be started as services since API 21
+    not exists(MethodAccess ma, Method m |
+      ma.getMethod() = m and
+      m.getDeclaringType().getASupertype*() instanceof TypeContext and
+      m.getName().matches(["start%Service%", "bindService%"]) and
+      this.asExpr() = ma.getArgument(0)
+    )
+    or
+    sinkNode(this, "pending-intent-sent")
+  }
+
+  override predicate hasState(DataFlow::FlowState state) { state = "MutablePendingIntent" }
+}
+
+/**
+ * Propagates taint from any tainted object to reads from its `PendingIntent`-typed fields.
+ */
+private class PendingIntentAsFieldAdditionalTaintStep extends ImplicitPendingIntentAdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(Field f |
+      f.getType() instanceof PendingIntent and
+      node1.(DataFlow::PostUpdateNode).getPreUpdateNode() =
+        DataFlow::getFieldQualifier(f.getAnAccess().(FieldWrite)) and
+      node2.asExpr().(FieldRead).getField() = f
```

Comment:
This step looks weird. It's propagating taint from `x` to `y.f` in
```
x.f = src
..
y.f
```
I'd expect either a step from `x` to `y` or a step from `src` to `y.f` instead, so if this somehow works it must implicitly rely on some other field-as-taint step.
Also, with the above comment cleared up, this is an instance of an ad-hoc field-write-to-read step without any link between `x` and `y`. Normally the data flow lib handles this for you when a path can be found from `x` to `y`. Can you elaborate on why this step is needed, i.e. why such an `x`-to-`y` path isn't expected to be found?

Reply:
A really good point.

I revisited this and the reason for its existence is [this test case](https://github.com/github/codeql/pull/6779/files/f103d4534042159018dd7fefcc91ec49d38b6dd8#diff-6a25d2186a74ca5e7a17bbeb8a9eb9f91197ffebf60852c4ad1a2c0be5697b3fR231), which is based on a real-world vulnerability found in AOSP.

After your reasoning I realized that what we were actually missing was modelling how the lifecycle of a `SliceProvider` works, i.e. `onCreateSliceProvider` is called implicitly when the provider is created, and then `onBindSlice` is called when a `Slice` is created using it.

So, I tested this by removing the existing field step and adding the following one (in `Slice.qll`):

```codeql
/**
 * An additional value step for modeling the lifecycle of a `SliceProvider`.
 * It connects the `PostUpdateNode` of any update done to the provider object in
 * `onCreateSliceProvider` to the instance parameter of `onBindSlice`.
 */
private class SliceProviderLifecycleStep extends AdditionalValueStep {
  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
    exists(Method onCreate, Method onBind, RefType declaringClass |
      declaringClass.getASupertype*() instanceof SliceProvider and
      onCreate.getDeclaringType() = declaringClass and
      onCreate.hasName("onCreateSliceProvider") and
      onBind.getDeclaringType() = declaringClass and
      onBind.hasName("onBindSlice")
    |
      node1.(DataFlow::PostUpdateNode).getEnclosingCallable() = onCreate and
      node1.getType() = declaringClass and
      node2.(DataFlow::InstanceParameterNode).getEnclosingCallable() = onBind
    )
  }
}
```

Does this look better?

Reply:
Yeah, that looks good. We really ought to have a nicer out-of-the-box way to specify that `node1` is an instance-access. I've put up a PR for that here: https://github.com/github/codeql/pull/7676. With that you can replace the two lines defining `node1` with
```
node1.(DataFlow::PostUpdateNode).getPreUpdate().(DataFlow::InstanceAccessNode).isOwnInstanceAccess() and
node1.getEnclosingCallable() = onCreate
```
That makes things a bit more precise and rules out any other nodes that might incidentally also have the same type `declaringClass`.

Reply:
Thanks so much! I changed it to use your wrapper and added it to the follow-up PR https://github.com/github/codeql/pull/7681.